src/Native/Unix/System.Native/pal_console.c

```diff
@@ -289,57 +320,76 @@ int32_t SystemNative_SetSignalForBreak(int32_t signalForBreak)
 {
     assert(signalForBreak == 0 || signalForBreak == 1);
 
-    struct termios current;
-    if (tcgetattr(STDIN_FILENO, &current) >= 0)
+    int rv = 0;
+
+    if (pthread_mutex_lock(&g_lock) == 0)
     {
-        IncorporateBreak(&current, signalForBreak);
-        if (tcsetattr(STDIN_FILENO, TCSANOW, &current) >= 0)
+        struct termios termios;
+        if (TcGetAttr(&termios, signalForBreak, g_hasInteractiveChildren))
         {
-            g_signalForBreak = signalForBreak;
-            return 1;
+            if (TcSetAttr(&termios))
+            {
+                g_signalForBreak = signalForBreak;
+                rv = 1;
+            }
         }
+        pthread_mutex_unlock(&g_lock);
     }
 
-    return 0;
+    return rv;
 }
 
 void ReinitializeConsole()
 {
-    // pal_signal.cpp calls this on SIGCONT/SIGCHLD.
-    // This can happen when SystemNative_InitializeConsole was not called.
-    // This gets called on a signal handler, we may only use async-signal-safe functions.
-
-    // If the process was suspended while reading, we need to
-    // re-initialize the console for the read, as the attributes
-    // previously set were likely overwritten.
-    if (g_readInProgress)
+    // Restores the state of the console after being suspended.
```

Comment:
This used to be called also for each SIGCHLD. Now we do this only on SIGCHLD when we are back at zero interactive children.